torch/csrc/jit/codegen/cuda/lower_vectorize_welford.cpp

```diff
@@ -0,0 +1,567 @@
+  //   float reciprocal = 1 / new_count;
+  //   for (i) {
+  //     welfordVectorized(..., new_count, reciprocal);
+  //   }
+  // }
+  void vectorizeWithLoopPredicate(WelfordOp* wop) {
+    kir::IfThenElse* wop_ite = scope_exprs_.back()->as<kir::IfThenElse>();
+    auto conditional = wop_ite->predicate()->value();
+
+    auto loop_ite = IrBuilder::create<kir::IfThenElse>(
+        IrBuilder::create<kir::Predicate>(conditional));
+
+    // Insert the IfThenElse and move the innermost loop inside it
+    registerInsertBeforeInnerMostLoop(loop_ite);
+    registerRemove(innermost_loop_, scope_of_innermost_loop_);
+    registerInsertBefore(nullptr, innermost_loop_, &loop_ite->thenBody());
+
+    // Change the scope of the innermost loop to the body of loop_ite
+    scope_of_innermost_loop_ = &loop_ite->thenBody();
+
+    auto vectorized_wop = applyVectorizeTransformation(wop, nullptr);
+
+    registerReplace(wop_ite, vectorized_wop, &innermost_loop_->body());
+  }
+
+  kir::VectorizedWelfordOp* applyVectorizeTransformation(
+      WelfordOp* wop,
+      Bool* pred) {
+    DataType data_type = wop->outAvg()->getDataType().value();
+    DataType index_type = wop->outN()->getDataType().value();
+
+    bool is_predicated = pred != nullptr;
+    if (!is_predicated) {
+      pred = GpuLower::current()->kernel()->trueVal();
+    }
+
+    // nvfuser_index_t new_count;
+    // new_count = hoisted_count + count_increment
+
+    auto new_count = defineScalar(index_type);
+
+    auto hoisted_count = hoistCount(wop->outN()->as<kir::TensorIndex>());
+
+    Int* count_increment = nullptr;
+    if (!is_predicated) {
+      count_increment = GpuLower::current()->kernel()->oneVal();
+    } else {
+      // count_increment = (int)pred;
+      count_increment = defineScalar(index_type)->as<Int>();
+      registerInsertBeforeInnerMostLoop(
+          IrBuilder::create<UnaryOp>(UnaryOpType::Cast, count_increment, pred));
+    }
+
+    registerInsertBeforeInnerMostLoop(IrBuilder::create<BinaryOp>(
+        BinaryOpType::Add, new_count, hoisted_count, count_increment));
+
+    // float new_count_float;
+    auto new_count_float = defineScalar(data_type);
+
+    // new_count_float = (float)new_count;
+    registerInsertBeforeInnerMostLoop(IrBuilder::create<UnaryOp>(
+        UnaryOpType::Cast, new_count_float, new_count));
+
+    // float reciprocal;
+    auto reciprocal = defineScalar(data_type);
+
+    auto reciprocal_expr = IrBuilder::create<BinaryOp>(
+        BinaryOpType::Div,
+        reciprocal,
+        GpuLower::current()->kernel()->oneVal(),
+        new_count_float);
+
+    // If not predicated, just set the reciprocal variable
+    // with the reciprocl expr. Otherwise, guard it with an if
+    // statement.
+    if (!is_predicated) {
+      registerInsertBeforeInnerMostLoop(reciprocal_expr);
+    } else {
+      // Initialize reciprocal as 0;
+      registerInsertBeforeInnerMostLoop(IrBuilder::create<UnaryOp>(
```

Comment:
Actually, it doesn't seem to work. `where` only accepts `Val*, so we would need to have `1/new_count_float` as a separate expression and use its output as an input to `where`, which means `1/new_count_float` would be executed unconditionally. Since `new_count_float` can be zero when the predicate is false, this would result in unsafe code. I'm going to leave it as is.